### PR TITLE
Part 3

### PR DESCRIPTION
**Pull request by anas14680**
Object Orient Structure for EHR Library

### CODE REVIEW

**Comment by anas14680** [comment]
Dear Professor,
@patrickkwang 
This is the way, I have designed my classes and parse data function. 

The patients data class does exactly what was intended. 

However, given that labs are not unique in the dataset, I created it as such that each attribute of the lab stores a list. For example, if a lab name is "URINALYSIS: RED BLOOD CELLS", the patient ID attribute of this lab's class will store a list of all the patients visiting this lab. If I follow this procedure, I can add 'Sick_patients' function as a method to the lab class.

Please let me know if this is fine, otherwise I can make a unique class for each lab data entry. If this is fine, I can proceed with tests, documentation and type hints.

On the bright side, my parse data function now has O(N) complexity.

File: Ehr_OOP.py [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+    """Store a patient data."""
+
+    def __init__(self, PatientID):
+        """Instantiate Patient Class."""
+        self.patientID = PatientID
+        self.gender = None
+        self.DOB = None
+        self.race = None
+
+    def __repr__(self):
```
Comment (patrickkwang) [inline Ehr_OOP.py]: `__repr__` typically produces a detailed string representation. For something this simple, I suspect that you want to overload `__str__` instead.

File: Ehr_OOP.py [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+
+from datetime import datetime
+
+
+class Patient:
+    """Store a patient data."""
+
+    def __init__(self, PatientID):
+        """Instantiate Patient Class."""
+        self.patientID = PatientID
```
Comment (patrickkwang) [inline Ehr_OOP.py]: PEP8 would have all variable and attribute names begin with a lowercase letter.

File: Ehr_OOP.py [hunk 3]
```diff
@@ -0,0 +1,97 @@
+"""Bio stats assignment #OOP Mohammad Anas."""
+
+
+from datetime import datetime
+
+
+class Patient:
+    """Store a patient data."""
+
+    def __init__(self, PatientID):
```
Comment (patrickkwang) [inline Ehr_OOP.py]: Add type hints throughout.

File: Ehr_OOP.py [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+        self.value = []
+        self.patientID = []
+
+    def __repr__(self):
+        """Lab Class Represenatation."""
+        return self.labname + " Class"
+
+    def store_info(self, lab_info):
+        """Extract Lab info."""
+        self.value.append(lab_info["LabValue"])
```
Comment (patrickkwang) [inline Ehr_OOP.py]: Why not add this stuff to the constructor (`__init__`)?

File: Ehr_OOP.py [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+        self.DOB = None
+        self.race = None
+
+    def __repr__(self):
+        """Patient Class Representation."""
+        return self.patientID + " Class"
+
+    def store_info(self, patientinfo):
+        """Extract info of a patient ID from text file."""
+        self.gender = patientinfo["PatientGender"]
```
Comment (patrickkwang) [inline Ehr_OOP.py]: Why not add this stuff to the constructor (`__init__`)?

File: Ehr_OOP.py [hunk 6]
```diff
@@ -0,0 +1,97 @@
...
+                    data_dict[information_dict["PatientID"]] = patient_class
+
+                elif pt_lb == "lab":
+                    lab_n = information_dict["LabName"]
+                    if lab_n not in data_dict:
+                        lab_class = Lab(lab_n)
+                        lab_class.store_info(information_dict)
+                        data_dict[lab_n] = lab_class
+                    else:
+                        data_dict[lab_n].store_info(information_dict)
```
Comment (patrickkwang) [inline Ehr_OOP.py]: Does this not overwrite the data already in there for another measurement with the same lab name?

File: Ehr_OOP.py [hunk 7]
```diff
@@ -0,0 +1,97 @@
...
+        """Extract Lab info."""
+        self.value.append(lab_info["LabValue"])
+        self.units.append(lab_info["LabUnits"])
+        self.patientID.append(lab_info["PatientID"])
+
+
+# redefine the function parse data that returns a class
+
+
+def parse_data(filename, pt_lb):
```
Comment (patrickkwang) [inline Ehr_OOP.py]: I'd probably just make separate `parse_patient_data` and `parse_lab_data` functions.

**Review by patrickkwang — CHANGES_REQUESTED** [review]
The patient stuff makes sense. I'm not so sure about `data_dict` for the labs, perhaps because I can't see yet how it will be used. If you are aiming for a `dict[lab_name, list[Lab]]`, that might serve for some of our applications. But I don't think that's what's happening currently.

File: Ehr_OOP.py [hunk 8]
```diff
@@ -0,0 +1,211 @@
...
+
+    def __init__(
+        self,
+        Labname: str,
+        units: str,
+        value: Union[str, float],
+        labDate: Union[str, datetime],
+    ) -> None:
+        """Instantiate Lab Class."""
+        self.labname: str = Labname  # O(1)
```
Comment (patrickkwang) [inline Ehr_OOP.py]:
```suggestion
        self.name: str = name  # O(1)
```
We already know that it's a lab...

File: Ehr_OOP.py [hunk 9]
```diff
@@ -0,0 +1,211 @@
...
+    def __init__(
+        self,
+        PatientID: str,
+        gender: str,
+        DOB: Union[str, datetime],
+        race: str,
+        labs: list[str],
+    ) -> None:
+        """Instantiate Patient Class."""
+        self.patientID: str = PatientID  # O(1)
```
Comment (patrickkwang) [inline Ehr_OOP.py]:
```suggestion
        self.id: str = id  # O(1)
```

File: Ehr_OOP.py [hunk 10]
```diff
@@ -0,0 +1,211 @@
...
+        self.gender: str = gender  # O(1)
+        self.DOB: datetime = datetime.strptime(DOB, "%Y-%m-%d %H:%M:%S.%f")  # O(2)
```
Comment (patrickkwang) [inline Ehr_OOP.py]:
```suggestion
        self.dob: datetime = datetime.strptime(dob, "%Y-%m-%d %H:%M:%S.%f")  # O(2)
```
PEP8 style guidance would have variable and attribute names start with a lowercase letter. We typically capitalize acronyms as if they were ordinary words.

File: Ehr_OOP.py [hunk 11]
```diff
@@ -0,0 +1,211 @@
+"""Bio stats assignment #OOP Mohammad Anas."""
```
Comment (patrickkwang) [inline Ehr_OOP.py:1]: We can replace the old stuff instead of making a new file.

File: tests.py [hunk 12]
```diff
@@ -1,196 +1,149 @@
...
+    answer = {"ANAS123": anas, "JON234": jon}
 
+    test = parse_data(labtest, pattest)
 
-# test num_older function
-def test_num_older():
-    """Test number older than a particular age."""
-    columns_only = "columns only.txt"
-    data_columns = parse_data(columns_only)
+    for i, j in zip(answer, test):
```
Comment (patrickkwang) [inline tests.py]: You should be able to compare these more concisely. You may want to implement `__eq__()` on your classes.

**Review by patrickkwang — CHANGES_REQUESTED** [review]

**Review by patrickkwang — APPROVED** [review]

**Pull request merged**